tests/framework/artifacts.py

```diff
@@ -0,0 +1,191 @@
+import platform
+
+from enum import Enum
+from pathlib import Path
+
+import boto3
+import botocore.client
+
+
+class ArtifactType(Enum):
+    """Supported artifact types."""
+
+    MICROVM = "microvm"
+    KERNEL = "kernel"
+    DISK = "disk"
+    SSH_KEY = "ssh_key"
+    MISC = "misc"
+
+
+class Artifact:
+    """A generic artifact manipulation class."""
+
+    def __init__(self, bucket, key, artifact_type=ArtifactType.MISC):
+        """Initialize bucket, key and type."""
+        self.bucket = bucket
+        self.key = key
+        self.local_folder = None
+        self._type = artifact_type
+
+    @property
+    def type(self):
+        """Return the artifact type."""
+        return self._type
+
+    def name(self):
+        """Return the artifact name."""
+        return Path(self.key).name
+
+    def local_dir(self):
+        """Return the directory containing the downloaded artifact."""
+        assert self.local_folder is not None
+        return "{}/{}/{}".format(
+            self.local_folder,
+            self.type.value,
+            platform.machine(),
+        )
+
+    def download(self, target_folder, force=False):
+        """Save the artifact in the folder specified target_path."""
+        self.local_folder = target_folder
+        Path(self.local_dir()).mkdir(parents=True, exist_ok=True)
+        if force or not os.path.exists(self.local_path()):
+            self.bucket.download_file(self.key, self.local_path())
+
+    def local_path(self):
+        """Return the local path where the file was downloaded."""
+        # The file path format: <target_folder>/<type>/<platform>/<name>
+        return "{}/{}".format(
+            self.local_dir(),
+            self.name()
+        )
+
+
+class DiskArtifact(Artifact):
+    """Specializes the generic artifact."""
+
+    def ssh_key(self):
+        """Return a ssh key artifact."""
+        # Replace extension.
+        key_file_path = str(Path(self.key).with_suffix('.id_rsa'))
+        return Artifact(self.bucket,
+                        key_file_path,
+                        artifact_type=ArtifactType.SSH_KEY)
+
+
+class ArtifactCollection:
+    """Provides easy access to different artifact types."""
+
+    MICROVM_CONFIG_EXTENSION = ".json"
+    MICROVM_KERNEL_EXTENSION = ".bin"
+    MICROVM_DISK_EXTENSION = ".ext4"
+
+    # S3 bucket structure.
+    ARTIFACTS_ROOT = 'ci-artifacts'
+    ARTIFACTS_DISKS = '/disks/' + platform.machine()
+    ARTIFACTS_KERNELS = '/kernels/' + platform.machine()
+    ARTIFACTS_MICROVMS = '/microvms'
+    ARTIFACTS_SNAPSHOTS = '/snapshots/' + platform.machine()
+
+    def __init__(
+        self,
+        bucket
+    ):
+        """Initialize S3 client."""
+        config = botocore.client.Config(signature_version=botocore.UNSIGNED)
+        # pylint: disable=E1101
+        # fixes "E1101: Instance of '' has no 'Bucket' member (no-member)""
+        self.bucket = boto3.resource('s3', config=config).Bucket(bucket)
+
+    def _fetch_artifacts(self,
```

Comment:
Do we plan on calling this with `artifact_root` different than `ArtifactCollection.ARTIFACTS_ROOT`?

Reply:
Nope, I've removed that parameter.